packages/devtools_app/lib/src/config_specific/drag_and_drop/drag_and_drop.dart

```diff
@@ -2,61 +2,169 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:flutter/gestures.dart';
 import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:provider/provider.dart';
 
 import '../../notifications.dart';
+import '../../utils.dart';
 import '_drag_and_drop_stub.dart'
     if (dart.library.html) '_drag_and_drop_web.dart'
     if (dart.library.io) '_drag_and_drop_desktop.dart';
 
-abstract class DragAndDrop extends StatefulWidget {
-  factory DragAndDrop({
-    @required Function(Map<String, dynamic> data) handleDrop,
-    @required Widget child,
-  }) {
-    return createDragAndDrop(
-      handleDrop: handleDrop,
-      child: child,
-    );
+abstract class DragAndDropManager {
+  factory DragAndDropManager({@required NotificationsState notifications}) {
+    final manager = createDragAndDropManager(notifications: notifications);
+    manager.init();
+    return manager;
+  }
+
+  DragAndDropManager.impl({@required this.notifications});
+
+  final NotificationsState notifications;
+
+  final _statesById = <String, DragAndDropState>{};
+
+  DragAndDropState activeState;
+
+  void init();
+
+  @mustCallSuper
+  void dispose() {
+    _statesById.clear();
+  }
+
+  void registerDragAndDrop(String id, DragAndDropState state) {
+    _statesById[id] = state;
+  }
+
+  void unregisterDragAndDrop(String id) {
+    _statesById.remove(id);
+  }
+
+  void dragOver(double x, double y) {
+    hitTestAndUpdateActiveId(x, y);
+    for (final state in _statesById.values) {
+      state.dragOver();
+    }
   }
 
-  const DragAndDrop.impl({@required this.handleDrop, @required this.child});
+  void dragLeave() {
+    for (final state in _statesById.values) {
+      state.dragLeave();
+    }
+  }
+
+  void drop() {
+    for (final state in _statesById.values) {
+      state.drop();
+    }
+  }
+
+  /// Performs a hit test to find the id for the active [DragAndDrop] widget at
+  /// the (x, y) coordinates, and updates the active state for all [DragAndDrop]
+  /// widgets accordingly.
+  void hitTestAndUpdateActiveId(double x, double y) {
+    final hitTestResult = HitTestResult();
+    RendererBinding.instance.hitTest(hitTestResult, Offset(x, y));
+
+    // Starting at bottom of [hitTestResult.path], look for [DragAndDrop] or
+    // [DragAndDropEventAbsorber] widgets. These widgets will be marked by a
+    // [RenderMetaData] target with [DragAndDropMetaData] metaData.
+    for (final result in hitTestResult.path) {
+      if (result.target is RenderMetaData) {
```

Comment:
you can avoid this as check in dart by writing
```dart
final target = result.target;
if (target is RenderMetaData) {
  // now target is known to be a RenderMetaData so not as check required.
  // Dart is picky about this because hitTestResult.target could possibly return a different value the second time it is called.
```
```

Reply:
done